packages/@glimmer/program/lib/program.ts

```diff
@@ -71,14 +74,26 @@ export class Heap implements CompileTimeHeap {
       this.table = table;
       this.offset = this.heap.length;
       this.handle = handle;
+      this.pageSize = this.heap.length;
     } else {
-      this.heap = new Uint16Array(0x100000);
+      this.heap = new Uint16Array(PAGE_SIZE);
       this.table = [];
     }
   }
 
   push(item: number): void {
     this.heap[this.offset++] = item;
+    this.sizeCheck();
```

Comment:
The size should be checked before an element is put on the heap, otherwise a size check should also be executed in the constructor in case of a `serializedHeap`.
This has also the advantage that we don't need to grow the heap in case we end on a page boundary :smiley: